# cigar_compare.pl

This program comapres cigars from reads that are tagged as mapping to multiple genomes.
  * If the alternative alignment has a higher total match value it will be treated as contamination and filtered out 
  * If the alternative alignment has a lower total match value it will be kept
  
When run in soft-clip mode (-s) it will also filter out reads softclipped > 20 at both ends and with a match < 30

# Input

This program accepts a bam file and runs it in samtools view using "samtools view -h $bam_file |". 

To filter a bam file for contamination and double soft-clipped reads run as: 

```
perl cigar_comapre.0.5.pl -s -b inputfile.bam | samtools sort - -o inputfile.filt.sc.bam
```

The program accepts as input a .bam file generated by running [tag_reads.py](https://github.com/bardin-lab/read_tagger).
Here, reads that map to an alternative genome as well as the primary genome are tagged as CR (ER..NR) and mates as DR.

Run in test mode (-t) to run on a [test dataset](files/test.bam)

Run as `perl cigar_comapre.0.5.pl -h` to see usage statement:

```
******** cigar_compare.pl ********
Usage: cigar_comapre.0.5.pl [options]
  --bam = specify bam file processed with 'tag_reads.py'
  --softclip = filter reads where both ends are softclipped > 20 and have a match < 30
  --test  = perform run on test dataset 'files/test.bam'
  --debug = run in debug mode
  --help = print this help message and exit

Nick Riddiford 2017
```

# Output

  * In normal mode (-b <bamfile>) a bed file will be produced showing the location of 'contamination' that has been filtered out, as well as the filtered bam file 
  * When combined with softclip mode (-b -s) reads will also be filtered as described in the usage statement

  * Run in debug mode (-d) to output cigar strings and total match scores (and decision) to ['.debug.log'](files/test.debug.log) for the first 10,000 reads, as well as a bed file ('mapped_not\_filtered.bed') showing the loactions of reads tagged as 'contamination' but kept in the alignment due to higher alignmnet scores to the primary genome
  * If debug and softclipped mode are combined (-d -s) it will output a bed file showing location of reads filtered on soft-clipping for the first 10,000 reads

